ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,765 @@
+    /// of records, i.e. `chunk_size` and the size of a record, i.e. `segment_size`.
+    /// ## Panics
+    /// Panics when `segment_size` is `None`.
+    fn allocate_vec(&mut self) {
+        // add 255 to round up
+        self.vec = vec![None; (self.chunk_size * self.segment_size.unwrap() + 255) >> 8];
+    }
+
+    /// `reset_offset` reallocates vector such that a smaller `RecordId` than `offset` can be stored
+    /// `new_offset` will be the new `offset` and the store and store elements with `RecordId` `new_offset` and larger
+    /// when `new_offset` is larger that the current `offset`, the function does nothing
+    fn reset_offset(&mut self, new_offset: RecordId, segment_length: usize) {
+        if new_offset < self.offset {
+            // metrics to count reallocations, disable-metrics flag will disable it globally
+            // right now, I don't pass the gate/step information which would be useful
+            // it would require to change the function signature and potentially clone gate even when feature is not enabled
+            metrics::increment_counter!(DZKP_BATCH_REALLOCATION_FRONT, RECORD => new_offset.to_string(), OFFSET => self.offset.to_string());
+
+            let mut new_offset = usize::from(new_offset);
+            // if segment_length is less than 256, redefine new offset such that we don't need to rearrange the existing segments
+            if 256 % segment_length == 0 {
+                // distance between new_offset and old offset is rounded up to a multiple of 256
+                new_offset = ((usize::from(self.offset) - new_offset + 255) >> 8) << 8;
+            }
+            let extension_length = ((usize::from(self.offset) - new_offset) >> 8) * segment_length;
+            // use existing vec and add enough space in front
+            self.vec.splice(0..0, repeat(None).take(extension_length));
+            self.offset = RecordId::from(new_offset);
+        }
+    }
+
+    /// `insert_segment` allows to include a new segment in `UnverifiedValuesStore`
+    ///
+    /// ## Panics
+    /// Panics when segments have different lengths across records
+    fn insert_segment(&mut self, record_id: RecordId, segment: Segment) {
+        // initialize `segment_size` when necessary
+        if self.segment_size.is_none() {
+            self.initialize_segment_size(segment.len());
+        }
+        // check segment size
+        debug_assert_eq!(segment.len(), self.segment_size.unwrap());
+        // allocate if vec is not allocated yet
+        if self.is_unallocated() {
+            self.allocate_vec();
+        }
+        // check offset
+        if record_id < self.offset {
+            // recover from wrong offset, expensive
+            self.reset_offset(record_id, segment.len());
+        }
+        let position_raw = usize::from(record_id) - usize::from(self.offset);
+        let length = segment.len();
+        let position_vec = (length * position_raw) >> 8;
```

Comment:
compilers are perfectly aware how to perform division by a power of two division I think. Your 32 byte assumption can be encoded as a constant and used everywhere

Reply:
didn't Ben Savage or Andy compare shift and divisions and shifts turned out to be more efficient?